Update network

simonvh opened the pull request:
This is an initial refactor of the network step. For now, major changes are to remove unnecessary code and to simplify. This can than serve as a base for adding new functionality. API is still compatible with current version, and the logic is also more or less the same.

Changes:
* Option to specify a list of TFs to use.

Bugfixes:
* Peaks that overlap promoter now get weight 1 instead of 0.
* Fixed issue with some annotation BED files that resulted in missing genes in the network.

simonvh commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
+    def unique_enhancers(self, binding_fname, chrom=None):
+        p = re.compile("[:-]")
+        logger.info("reading enhancers")
+        enhancers = pd.read_table(binding_fname, usecols=["enhancer"])["enhancer"]
```
This is **really** inefficient for large network files. However, the information is not present in an other file. Maybe the network should be stored in a different format, not in a flat text file. Could be a hdf5 file, with seperate tables per chromosome, and one table that stores all enhancers. Or something like that. You lose the benefit of flat text, but hopefully can store data more efficiently, and also enable more efficient reading.

simonvh replied:
Another advantage of doing certain things per chromosome, is that we can do stuff in chunks parallel, or serial (in case of a machine with lower memory than cn45)

Maarten-vd-Sande replied:
One thing to consider is maybe sqlite r-tree. It's the perfect datastructure for this (linking proximalish peaks with tss) I think, and should be fast / dynamic. I implemented it for peaksql, and I could try a bit if it's (much) faster. The advantage is that we can put everything on disk, so the memory usage should be very low, and sqlite in general is fast. Big disadvantage is maintainability and perhaps portability between systems..

siebrenf commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
-        bpd = pd.DataFrame(bp)
-        bpd = bpd.rename(columns={0: "binding"})
-        bpd["prob"] = minmax_scale(rankdata(bpd["binding"], method="dense"))
+        if tfs is None:
+            package_dir = os.path.dirname(ananse.__file__)
+            tffile = os.path.join(package_dir, "db", "tfs.txt")
+            tfs = pd.read_csv(tffile, header=None)[0].tolist()
```
- If `tfs` is specified, it must currently be a list, I think a file might be nicer.
 - No option to pass tfs over the CLI.
 - No option to _not_ filter by TFs

simonvh replied:
I added this functionality for now only to use via the API. It makes generating networks for a limited set of TFs *much* faster. Later on, this would also be exposed via the CLI. 
I actually like it to be a list for this function. The CLI could actually also use a file.
Given that the point of ANANSE is predicted TF-target interactions, what would be the use case of NOT filtering? :) Actually, what it mainly does now is removing wrong IDs from the motif file that cannot be resolved.

siebrenf commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
 
-            logger.info("Create network")
-            self.create_promoter_network(ddf, prom, fin_expression, outfile)
+        expression["key"] = 0
+        tfs["key"] = 0
```
unused (?)

simonvh replied:
Yeah, leftover. Not sure why they are in here. It has something to do with generating the df after that. Will check it out.

siebrenf commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
 
+        # Use a version of the binding network, either promoter-based, enhancer-based
+        # or both.
+        if self.include_promoter or self.include_enhancer:
```
this if else structure contains the following three options:
- if only promoters/enhancers are used
- if `fun_expression` is None.
- if both are used

the first two are currently merged. Perhaps it is clearer if they are separated? That way we can also skip `create_expression_network` if the input is `None` anyway

simonvh replied:
> A couple of remarks, not all related to this PR, but more to the network.py file in general. The PR is nice
> 
> * There is a lot of stuff being written and read to disk, which often seems unnecessary. Perhaps switching from pybedtools to pyranges might solve some of these issues.

Yes, we should see where this is not necessary. One thing I would like to keep in mind is the memory tradeoff.. it should also be usable on a machine with reasonable RAM (say 16-32G)

> * I never worked with dask, not sure how some stuff works. Are some computations "futures"? Which you do later?

Yes! It is now written that most the computation is actually performed on the `compute()` call. This means we can later manage this by setting up a dask.distributed client that manages workers and memory consumption.

> * Variable and function names are often unclear. This, in combination with the lack of comments and docstrings, make the code quite hard to read sometimes, even though the code is really neat. The stuff that happens in network isn't overly complicated, but every time I look at a function I have no clue what it happens until I read it carefully. Adding docstrings to functions with an example input and output would make reading the code much easier.

I know.. I've started adding comments and updating variable names, but there's still a lot to do there!

siebrenf commented on ananse/network.py:
```diff
@@ -213,408 +180,117 @@ def distance_weight(self, include_promoter=False, include_enhancer=True, alpha=1
+            alpha = padding / 10
+
+        if promoter > padding:
+            raise ValueError("promoter region is > region to use")
```
I have no idea what this means, and why this is important. Could you add docstrings to this function?

siebrenf commented on ananse/network.py:
```diff
@@ -196,13 +145,31 @@ def get_gene_dataframe(self, peak_bed, up=100000, down=100000):
         return p
 
-    def distance_weight(self, include_promoter=False, include_enhancer=True, alpha=1e4, padding=100000, keep1=5000, remove=2000):
+    def distance_weight(
```
Please add docstrings explaining the variables and weights 1-3?

siebrenf reviewed (commented):
First of all, I'm really happy to see the code become so much more manageable!!

There are still some non-descriptive variables, but those are fine if there would be docstrings (I marked two critical functions that would really benefit). 

The only question I have from a coding perspective, is the use of `tfs`, which happens in multiple places, cannot be specified from the CLI side and seems questionably useful since there are filter steps in ANANSE binding (which could be expanded there for lowered compuational costs?). I would however ignore this point for now.

Maarten-vd-Sande reviewed (commented):
A couple of remarks, not all related to this PR, but more to the network.py file in general. The PR is nice

- There is a lot of stuff being written and read to disk, which often seems unnecessary. Perhaps switching from pybedtools to pyranges might solve some of these issues.
- I never worked with dask, not sure how some stuff works. Are some computations "futures"? Which you do later?
- Variable and function names are often unclear. This, in combination with the lack of comments and docstrings, make the code quite hard to read sometimes, even though the code is really neat. The stuff that happens in network isn't overly complicated, but every time I look at a function I have no clue what it happens until I read it carefully. Adding docstrings to functions with an example input and output would make reading the code much easier.

Maarten-vd-Sande commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
+        enhancers = enhancers.unique()
+        logger.info("writing temporary enhancers")
+        f = NamedTemporaryFile(mode="w+", dir=mytmpdir(), delete=False)
+        for e in enhancers:
+            print("{}\t{}\t{}".format(*re.split(p, e)), file=f)
+
```
Also I think you could do this with pandas `df[col].str.split`

simonvh replied:
For now this is not the most inefficient step in the whole process. I actually don't know if the pandas version would be faster.. should check.

siebrenf commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
+            if fin_expression is not None:
+                with ProgressBar():
+                    result = df_expression.join(df_binding)
+                    result = result.compute()
```
dask only runs the actual code once a `compute` function is called. `dask.diagnostics.ProgressBar` tracks the computation (which is not prettied up by loguru unfortunately, but useful)

Maarten-vd-Sande commented on ananse/network.py:
```diff
@@ -196,13 +145,31 @@ def get_gene_dataframe(self, peak_bed, up=100000, down=100000):
+        ).set_index("dist")
+        p = p.join(weight, on="dist")
 
+        p.gene = [i.upper() for i in list(p.gene)]
```
I'm btw not a fan of just changing gene names into upper or lower. But this has to be changed simultaneously in binding as well

siebrenf replied:
in binding, TF factors were (and still are) also converted to upper case

simonvh replied:
No indeed! This is left-over. We should not need to touch gene names in this manner.

Maarten-vd-Sande commented on ananse/network.py:
```diff
@@ -73,89 +75,36 @@ def __init__(self, ncore=1, genome="hg38", gene_bed=None, include_promoter=False
+    ):
         # all overlap Enh-TSS(100000-tss-100000) pair with distance
+        logger.info("relating enhancers to genes")
         peaks = BedTool(peak_bed)
```
p.s. another check if there are only unique genes in here? In e.g. xt10 there are multiple of the same gene

https://raw.githubusercontent.com/vanheeringen-lab/ANANSE/master/ananse/db/xt10.0.gene.bed (e.g. efna3)

Not sure what the effects of that is downstream

Maarten-vd-Sande commented:
Was it a bug that promoters were zero btw?
![image](https://user-images.githubusercontent.com/15323399/108395174-cfbbc080-7215-11eb-978b-bfddf232cb9f.png)

Seems intentional

Maarten-vd-Sande commented on ananse/network.py:
```diff
@@ -213,408 +180,117 @@ def distance_weight(self, include_promoter=False, include_enhancer=True, alpha=1
-        weight = dd.read_csv(weight)
-        f_table = f_table.merge(weight, how="left", on="dist")
-        f_table["sum_dist_weight"] = f_table["binding"] * f_table["weight"]
+        weight = pd.concat([weight1, weight2, weight3])
```
why are this dataframes? Doesn't a single list just work? The `dist` column == index?

simonvh commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
-        bpd = pd.DataFrame(bp)
-        bpd = bpd.rename(columns={0: "binding"})
-        bpd["prob"] = minmax_scale(rankdata(bpd["binding"], method="dense"))
+        if tfs is None:
+            package_dir = os.path.dirname(ananse.__file__)
+            tffile = os.path.join(package_dir, "db", "tfs.txt")
+            tfs = pd.read_csv(tffile, header=None)[0].tolist()
```
We should :).

simonvh commented on ananse/network.py:
```diff
@@ -213,408 +180,117 @@ def distance_weight(self, include_promoter=False, include_enhancer=True, alpha=1
+            raise ValueError("promoter region is > region to use")
+
+        # Get list of unique enhancers from the binding file
+        # This is relatively slow for a large file. May need some optimization.
+        enhancer_bed = self.unique_enhancers(binding_fname)
+
```
The mean reason we offload to bedtools is to determine distances of enhancers to genes. If we can use an efficient pyranges function there that would be great.

simonvh commented on ananse/network.py:
```diff
@@ -73,89 +75,36 @@ def __init__(self, ncore=1, genome="hg38", gene_bed=None, include_promoter=False
+        peak_bed,
+        up=100000,
+        down=100000,
+        alpha=1e4,
```
I didn't know about this! :shocked:

simonvh commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
+        network["tf_target"] = network["tf"] + "_" + network["target"]
+        network = network.set_index("tf_target").drop(columns=["tf", "target"])
+        logger.info("creating expression dataframe")
+        network = dd.from_pandas(network, npartitions=30)
```
It creates a dask dataframe with the Cartesian product of all TFs and target genes.

simonvh commented on ananse/network.py:
```diff
@@ -73,89 +75,36 @@ def __init__(self, ncore=1, genome="hg38", gene_bed=None, include_promoter=False
-                self.gene_bed =  os.path.join(package_dir, "db", "hg38.genes.bed")
+                self.gene_bed = os.path.join(package_dir, "db", "hg38.genes.bed")
             else:
                 self.gene_bed = gene_bed
         elif self.genome == "hg19":
             if gene_bed is None:
-                self.gene_bed =  os.path.join(package_dir, "db", "hg19.genes.bed")
+                self.gene_bed = os.path.join(package_dir, "db", "hg19.genes.bed")
             else:
-                self.gene_bed = gene_bed            
+                self.gene_bed = gene_bed
         else:
             if gene_bed is None:
                 raise TypeError("Please provide a gene bed file with -a argument.")
             else:
                 self.gene_bed = gene_bed
-        
+
```
I would actually like the human-based analysis to be as hassle-free as possible. Maybe we could use the genomepy annotation file by default? And then supply the option to provide another annotation file?

siebrenf replied:
We could let ANANSE accept a genome name, then download the genome and annotation with genomepy if not found (returning a warning if no annotation can be downloaded). 

In this function we can then allow a custom annotation input OR a filter mechanism as you mentioned above (both might be overkill).

simonvh commented on ananse/network.py:
```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
-        tfs = pd.read_csv(tffile, header=None)[0].tolist()
-
-        tfs = expression[expression.gene.isin(tfs)]
+        tfs = expression[expression.target.isin(tfs)]
```
This info doesn't come from binding. This is based on the RNA-seq data, to generate a "network" where we use TF expression and target gene expression.

simonvh commented:
@siebrenf @Maarten-vd-Sande I think I addressed most of your comments. There are some outstanding issues, but I think these may be fixed in a later round of polishing.

siebrenf commented on ananse/network.py:
```diff
@@ -213,408 +180,117 @@ def distance_weight(self, include_promoter=False, include_enhancer=True, alpha=1
+        logger.info("Reading binding file...")
+        ddf = dd.read_csv(
+            binding_fname,
+            sep="\t",
```
this is double now :)
```
        ddf = dd.read_csv(
            binding_fname,
            sep="\t",
            usecols=["factor", "enhancer", "binding"],
        )[["factor", "enhancer", "binding"]]
```

Maarten-vd-Sande commented on ananse/network.py:
```diff
@@ -73,728 +73,512 @@ def __init__(self, ncore=1, genome="hg38", gene_bed=None, include_promoter=False
-            logger.info("Read data")
-            ddf = dd.read_csv(binding, sep="\t")[["factor", "enhancer", "binding"]]
+        # Merge TF and target gene expression information
+        network = expression.merge(tfs, how="outer")
```
I think this needs to be:

```
network = expression.merge(tfs, how="outer", left_on="target", right_on="tf")
```
